scipy/optimize/tests/test_milp.py

```diff
@@ -0,0 +1,248 @@
+"""
+Unit test for Mixed Integer Linear Programming
+"""
+
+import re
+import numpy as np
+import pytest
+from numpy.testing import assert_allclose, assert_array_equal
+from .test_linprog import magic_square
+from scipy.optimize import milp, Bounds, LinearConstraint
+
+
+def test_milp_iv():
+
+    message = "`c` must be a one-dimensional array"
+    with pytest.raises(ValueError, match=message):
+        milp(np.zeros((3, 4)))
+
+    message = "`bounds` must be an instance of..."
+    with pytest.raises(ValueError, match=message):
+        milp(1, bounds=10)
+
+    message = "`constraints` must be a sequence of..."
+    with pytest.raises(ValueError, match=message):
+        milp(1, constraints=10)
+
+    message = "The shape of `A` must be (len(b_l), len(c))."
+    with pytest.raises(ValueError, match=re.escape(message)):
+        milp(np.zeros(3), constraints=([[1, 2, 3]], [2, 3], [2, 3]))
+    with pytest.raises(ValueError, match=re.escape(message)):
+        milp(np.zeros(3), constraints=([[1, 2]], [2], [2]))
+
+    message = ("`integrality` must contain integers 0-3 and be broadcastable "
+               "to `c.shape`.")
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], integrality=[1, 2])
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], integrality=[1, 5, 3])
+
+    message = "`lb`, `ub`, and `keep_feasible` must be broadcastable."
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], bounds=([1, 2], [3, 4, 5]))
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], bounds=([1, 2, 3], [4, 5]))
+
+    message = "`bounds.lb` and `bounds.ub` must contain reals and..."
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], bounds=([1, 2], [3, 4]))
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], bounds=([1, 2, 3], ["3+4", 4, 5]))
+    with pytest.raises(ValueError, match=message):
+        milp([1, 2, 3], bounds=([1, 2, 3], [set(), 4, 5]))
+
+
+def test_milp_options(capsys):
+    message = "Unrecognized options detected: {'ekki'}..."
+    options = {'ekki': True}
+    with pytest.warns(RuntimeWarning, match=message):
+        milp(1, options=options)
+
+    A, b, c, numbers, M = magic_square(3)
+    options = {"disp": True, "presolve": False, "time_limit": 0.05}
+    res = milp(c=c, constraints=(A, b, b), bounds=(0, 1), integrality=1,
+               options=options)
+
+    # This shows that captured.out tests work
+    # Need to pass this test when the following print statement is _removed_.
+    print("Time Limit Reached, Presolve is switched off")
```

Comment:
Do I read this correctly, you will remove this statement at some point?

Reply:
Yes. I added this test because I have noticed that on my machine, `linprog` doesn't ever print to the console in my IDE; it prints to the console window my IDE was opened from. This test confirms that something is not quite right about the way that `milp` (and `linprog`) route output from HiGHS.

I think we should remove the `print` statement and `xfail` the test for now. I'll open a separate issue about it, and when it gets fixed there, we can remove the `xfail` mark.

Reply:
You could `xfail` using `strict=True`. Then it would raise when suddenly the test passes. A good way to keep track of issues in the test suite.

Reply:
Fixed in next commit.